fix: pivot table and multi-statement support

rschmukler opened the pull request:
This commit introduces support for pivot tables by adding support for multiple statements inside a prepared statement. This is because a pivot table will automatically create an anonymous type for the result before running.

This commit includes several refactors to help cleanup the code.

1) Updates the duckdb-ffi function defenitions by changing `:pointer` to `by-value :type` defenitions. This allows the use of resource tracking on the structs to call the cleanup functions, as well as allowing the structs to be passed as `:pointer` when approprirate (avoiding a lot of the manual `Pointer.` construction) via the auto casting of `->pointer`.

2) Get rid of the multiple types for `PrepStatement` arities and instead just defenining multiple `IFn` impls for the various arities.

3) Get rid of the `param-types` vector which wasn't being used in the current version of param binding.

4) Get rid of the need to destroy resources manually and instead rely on the resource tracking.

5) Introduce a `->error` fn helper for generating error messages from c-string ptrs since we had a lot of code that was doing the required ceremony inline.

6) Introduce a `-execute-stmt!` helper that handles resource binding and cleanup to make it easier to understand the logic inside `prepare!`

One thing of note is that the `AutoClosable` implementation doesn't explicitly free the result anymore. We now instead wait for the GC to get rid of it. If you prefer that I also release them on `AutoClosable` it's relatively easy to support both. Right now the AutoClosable impl is a `no-op` effectively.

rschmukler committed:
fix: pivot table and multi-statement support

This commit introduces support for pivot tables by adding support for
multiple statements inside a prepared statement. This is because a pivot
table will automatically create an anonymous type for the result before
running.

This commit includes several refactors to help cleanup the code.

1) Updates the duckdb-ffi function defenitions by changing `:pointer` to
`by-value :type` defenitions. This allows the use of resource tracking
on the structs to call the cleanup functions, as well as allowing the
structs to be passed as `:pointer` when approprirate (avoiding a lot of
the manual `Pointer.` construction) via the auto casting of `->pointer`.

2) Get rid of the multiple types for `PrepStatement` arities and instead
just defenining multiple `IFn` impls for the various arities.

3) Get rid of the `param-types` vector which wasn't being used in the
current version of param binding.

4) Get rid of the need to destroy resources manually and instead rely on the resource tracking.

5) Introduce a `->error` fn helper for generating error messages from
c-string ptrs since we had a lot of code that was doing the required
ceremony inline.

6) Introduce a `-execute-stmt!` helper that handles resource binding and
cleanup to make it easier to understand the logic inside `prepare!`

cnuernber reviewed (changes requested):
The only thing I see is that in your method to create a new struct you rely on the resource system to release things in reverse order of how they are registered.  If we are using the stack-based pathway then that will happen but if we are using the JVM GC then we do not get that guarantee - thus it is possible the struct is 'free'd before the destroy function is called and if the system is under load (or run with debug settings in the C runtime) the values in the struct could be scambled by the time the destroy function is called.

My suggestion is to always allocate the struct with resource-type nil - this disables auto free and manually free it after the destroy method is called if the destroy method is provided.   Then we are sure the destroy method - if provided - is called with valid values.

Aside from that I think this is fine.  It does disable the manual with-open pathway to free the prep statement but we can see if anyone complains or perhaps file a second issue to fix that later - I agree it is easy to address later but that is why I used the delay-based system so that we avoid double free and allow both a manual pathway and automatic.

rschmukler committed:
chore: update nix flake build

Updates the nix flake build to handle the new multi-output format of the
duckdb library.

cnuernber reviewed (approved):
Great addition - Thanks Ryan!!
